crypto/x509/x509_lu.c

```diff
@@ -118,6 +118,23 @@ int X509_LOOKUP_by_alias(X509_LOOKUP *ctx, X509_LOOKUP_TYPE type,
     return ctx->method->get_by_alias(ctx, type, str, len, ret);
 }
 
+int X509_LOOKUP_set0_method_data(X509_LOOKUP *ctx, void *data)
+{
+    ctx->method_data = data;
+    return 1;
+}
```

Comment:
Elsewhere it was suggested that "set0" semantic implies kind of ownership transfer, in sense that callee is freed from responsibility to manage the object, most notably free it. So I suppose question here is if this is right?

Reply:
That interpretation is incorrect, mind pointing me to the discussion?

The interpretation of `set0` is that the library will share the data with the caller, and that the caller should be careful what they do with the provided data (for example, freeing the method data before the X509_LOOKUP has been freed would be perfectly stupid).

The interpretation of `set1` is that the application *will make a copy* (or simulate doing so by incrementing a reference counter) of the provided data.  The caller is then free to free its reference (for reference counted stuff, that decrements the counter, which effectively transfers ownership of the object to the library).

Reply:
*ahem* ... of course, if the caller simply drops the pointer to the object, `set0` works as transfer of ownership as well...

Reply:
I suppose @dot-asm is talking about the need to free the old contents of ctx->method_data if there was any. And yes, that should be done unless changing the method_data if previously set is completely unwanted, but then the function should return error in case it was already set.

Reply:
No.  The method data isn't for the consumption of the library, it is for the consumption of the method functions only.  The library know nothing about that pointer *and shouldn't* know anything about that pointer.

Reply:
See `crypto/x509/by_dir.c` to see how a built in `X509_LOOKUP` method uses it.

Reply:
set0 normally implies that responsibility for management of the memory passes to the object, e.g. so that when the object is freed, so is any memory passed in via set0. If it doesn't do that, then its not a set0 IMO.

Reply:
In that case, we have some confusing definitions of what `set0` should do.  Frankly, I cannot remember the discussions surrounding that construct any more.  Point is, if @mattcaswell's interpretation is the correct one, then we do have a problem, because there are `set0` functions already in the library with exactly the semantics of this one, i.e. it passes around user data (or in this case, method data).  For example, see `DH_meth_set0_app_data`

Reply:
I can give more examples:

`EVP_PKEY_CTX_set0_keygen_info`
`PKCS7_set0_type_other`
`X509_STORE_CTX_set0_*` (except `X509_STORE_CTX_set0_param`)
`GENERAL_NAME_set0_*`


But I do see functions that do what @mattcaswell says as well, for example `NAMING_AUTHORITY_set0_*`, `SSL_set0_CA_list`, `SSL_set0_rbio`, `SSL_set0_wbio`, ...

So it seems like we aren't consistent, and I suspect that it's because we have different understandings of what `set0` is supposed to mean.

Reply:
The existing functions cannot be changed but new incorrectly named one should not be introduced. So I would vote for naming the function X509_LOOKUP_set_method_data without the 0.

Reply:
Sure, I can do that...  side note: we *really* should document the `set` / `set0` / `set1` pattern, or these kinds of discussions will come again and again, and the current examples will keep on confusing us.

Reply:
But then, shouldn't the getters be named similarly?  What's the meaning of `get0` in this scenario?

Reply:
What kind of gets up my nose is to have a setter called `WHATEVER_set_something` and that the corresponding getter would be called `WHATEVER_get0_something`, there's a lack of orthogonality there that rubs me the wrong way.  But, we can also note that there's no difference at all between `get` and `get0`, except the latter is a bit more explicit.

So I can rename this setters to use `set` and the two getters to use `get` instead of `get0`.  Does that work with everyone?  Is that the pattern we want for these kinds of setters / getters?